// @solvro/config/dist/adonis-MGLA5SOL.js
import {
  node
} from "./chunk-AQ5RXFKK.js";
import {
  imports
} from "./chunk-VSUL3MI5.js";
import "./chunk-HM3MY7J3.js";
import {
  init_esm_shims
} from "./chunk-J5OUMHDO.js";

// src/eslint/presets/adonis.ts
init_esm_shims();
import { configApp } from "@adonisjs/eslint-config";
var adonisPreset = () => {
  const builtinAdonisConfig = configApp();
  return [
    ...builtinAdonisConfig,
    ...node(),
    ...imports(),
    {
      rules: {
        "@typescript-eslint/naming-convention": [
          "error",
          {
            selector: ["enum", "enumMember", "class", "interface", "typeLike"],
            format: ["PascalCase"],
            leadingUnderscore: "allow",
            trailingUnderscore: "allow"
          },
          {
            selector: [
              "classProperty",
              "classMethod",
              "method",
              "variableLike"
            ],
            format: ["camelCase"],
            leadingUnderscore: "allow",
            trailingUnderscore: "allow"
          },
          {
            selector: "variable",
            format: ["camelCase", "UPPER_CASE", "PascalCase"],
            leadingUnderscore: "allow",
            trailingUnderscore: "allow"
          }
        ]
      }
    }
  ];
};
export {
  adonisPreset
};
