// @solvro/config/dist/chunk-VSUL3MI5.js
import {
  default2,
  default3
} from "./chunk-HM3MY7J3.js";
import {
  init_esm_shims
} from "./chunk-J5OUMHDO.js";

// src/eslint/configs/imports.ts
init_esm_shims();
var forbiddenUiLibraries = [
  "@headlessui/react",
  "@mui/material",
  "@chakra-ui/react",
  "@chakra-ui/core",
  "@nextui-org/react",
  "react-bootstrap",
  "antd"
];
function imports({
  forbidDefaultExport = true
} = {}) {
  const config = [
    default3.flatConfigs.typescript,
    {
      name: "solvro/imports/rules",
      plugins: {
        antfu: default2
      },
      rules: {
        "antfu/import-dedupe": "error",
        "antfu/no-import-dist": "error",
        "antfu/no-import-node-modules-by-path": "error",
        ...default3.flatConfigs.recommended.rules,
        "import/no-dynamic-require": "warn",
        "import/no-unresolved": "off",
        "import/consistent-type-specifier-style": "warn",
        "@typescript-eslint/no-restricted-imports": [
          "error",
          {
            paths: [
              {
                name: "axios",
                message: "Please use fetch instead"
              },
              ...forbiddenUiLibraries.map((library) => ({
                name: library,
                message: `Please use ui.shadcn.com components instead.`
              }))
            ]
          }
        ]
      }
    }
  ];
  if (forbidDefaultExport) {
    config.push(
      {
        rules: { "import/no-default-export": "error" }
      },
      {
        files: [
          "tsup.config.*",
          "eslint.config.*",
          ".commitlintrc.*",
          "knip.*",
          "next.config.*",
          "commitlint.config.*",
          "vite.config.*",
          ".releaserc.*",
          "vitest.config.*",
          "postcss.config.*",
          "playwright.config.*"
        ],
        rules: {
          "import/no-default-export": "off"
        }
      }
    );
  }
  return config;
}

export {
  imports
};
