// @darraghor/eslint-plugin-nestjs-typed/dist/rules/injectablesShouldBeProvided/injectableShouldBeProvided.js
import { createRule } from "../../utils/createRule.js";
// import FileEnumeratorWrapper from "../../utils/files/fileEnumerationWrapper";
import NestProvidedInjectableMapper from "../../utils/nestModules/nestProvidedInjectableMapper.js";
import { typedTokenHelpers } from "../../utils/typedTokenHelpers.js";
import FileEnumeratorWrapper from "../../utils/files/customFileEnumeratorWrapper.js";
let listOfPotentialNestModuleFiles;
let nestModuleMap;
const findModuleMappingsContaining = (className, propertyName, nestModuleMap) => {
    // eslint-disable-next-line @typescript-eslint/naming-convention, @typescript-eslint/no-unused-vars
    const modules = [...nestModuleMap.entries()].filter(([_, entry]) => {
        return entry[propertyName].has(className);
    });
    return modules;
};
function convertNestModuleMapToString(nestModules) {
    // const uniqueValues = new Set<string>();
    const vals = [];
    nestModules.forEach((module) => {
        // Add all paths
        vals.push(module[0]);
    });
    // Convert the Set to an array and join with commas
    // return Array.from(uniqueValues).join(", ");
    return vals.join(", ");
}
const checkNode = (
// super fragile types but whatevs
node, decoratorName, propertyName, messageId, context) => {
    if (NestProvidedInjectableMapper.isNestInjectableThatIsNeverProvided(node)) {
        return;
    }
    if (typedTokenHelpers.nodeHasDecoratorsNamed(node, [decoratorName])) {
        const name = node.id?.name;
        if (!name) {
            return;
        }
        const foundModuleMaps = findModuleMappingsContaining(name, propertyName, nestModuleMap);
        if (foundModuleMaps.length === 1) {
            return;
        }
        // couldn't find map so error
        context.report({
            node: node,
            messageId: messageId,
            data: {
                numberOfReferences: foundModuleMaps.length,
                refLocations: convertNestModuleMapToString(foundModuleMaps),
                name: name,
            },
        });
    }
};
function initializeModuleMappings(sourcePath, filterFromPaths, context) {
    const mappedSourceDirectory = NestProvidedInjectableMapper.detectDirectoryToScanForFiles(sourcePath, process.cwd());
    listOfPotentialNestModuleFiles = FileEnumeratorWrapper.enumerateFiles({
        sourceGlobs: mappedSourceDirectory,
        extensions: [".ts"],
        filterFromPaths,
    });
    nestModuleMap = NestProvidedInjectableMapper.parseFileList(listOfPotentialNestModuleFiles, context);
}
const defaultOptions = [
    {
        src: ["src/**/*.ts"],
        filterFromPaths: ["dist", "node_modules", ".test.", ".spec."],
    },
];
const rule = createRule({
    name: "injectable-should-be-provided",
    meta: {
        docs: {
            description: "Public api methods should have documentation",
        },
        messages: {
            injectableInModule: `Expected Injectable "{{name}}" to be referenced in 1 module's providers array, but found {{numberOfReferences}} references ({{refLocations}}). If you added it already but this error still shows in your editor, please change one character in the injectable file to poke your eslint plugin.`,
            controllersInModule: `Expected Controller "{{name}}" to be referenced in 1 module's controllers array, but found {{numberOfReferences}} references ({{refLocations}}). If you added it already but this error still shows in your editor, please change one character in the controller file to poke your eslint plugin.`,
        },
        schema: [
            {
                type: "object",
                properties: {
                    src: {
                        description: "files/paths to be analyzed (only for provided injectable or controller)",
                        type: "array",
                        minItems: 1,
                        items: {
                            type: "string",
                            minLength: 1,
                        },
                    },
                    filterFromPaths: {
                        description: "strings to exclude from checks (only for provided injectable or controller)",
                        type: "array",
                        minItems: 1,
                        items: {
                            type: "string",
                            minLength: 1,
                        },
                    },
                },
            },
        ],
        type: "problem",
    },
    defaultOptions: defaultOptions,
    create(contextWithoutDefaults) {
        // @ts-ignore
        const context = contextWithoutDefaults.options &&
            contextWithoutDefaults.options.length > 0
            ? contextWithoutDefaults
            : // only apply the defaults when the user provides no config
                Object.setPrototypeOf({
                    options: defaultOptions,
                }, contextWithoutDefaults);
        const { src, filterFromPaths } = context.options[0] || {};
        if (nestModuleMap === undefined || nestModuleMap.size === 0) {
            initializeModuleMappings(src[0], filterFromPaths, context);
        }
        return {
            ClassDeclaration(node) {
                checkNode(node, "Injectable", "providers", "injectableInModule", context);
                checkNode(node, "Controller", "controllers", "controllersInModule", context);
            },
            "Program:exit"() {
                // map the source to a mapping thing
                // if not undefined set it to the mapping set
                const mappedProvidedInjectables = NestProvidedInjectableMapper.mapAllProvidedInjectablesInModuleOrProviderFile(context.sourceCode.ast, context.filename);
                if (mappedProvidedInjectables !== null) {
                    nestModuleMap.set(mappedProvidedInjectables[0], mappedProvidedInjectables[1]);
                }
            },
        };
    },
});
export default rule;
