// @darraghor/eslint-plugin-nestjs-typed/dist/utils/typedTokenHelpers.js
import { AST_NODE_TYPES, TSESTree } from "@typescript-eslint/utils";
import { parse } from "@typescript-eslint/parser";
import ts from "typescript";
import * as tsutilsImp from "ts-api-utils";
const tsutils = tsutilsImp;
export const typedTokenHelpers = {
    decoratorsThatCouldMeanTheDevIsValidatingAnArray: [
        "IsArray",
        "ArrayMinSize",
        "ArrayMinSize",
        "ArrayContains",
        "ArrayNotContains",
        "ArrayNotEmpty",
        "ArrayUnique",
    ],
    isTypeArrayTypeOrUnionOfArrayTypes(node, parserService, checker) {
        if (node.typeAnnotation?.typeAnnotation
            .type === TSESTree.AST_NODE_TYPES.TSArrayType) {
            return true;
        }
        const nodeType = this.getNodeType(node, parserService, checker);
        if (checker.isArrayType(nodeType)) {
            return true;
        }
        for (const t of tsutils.unionTypeParts(nodeType)) {
            if (!checker.isArrayType(t)) {
                return false;
            }
        }
        return true;
    },
    getNodeType(node, parserService, checker) {
        const tsNode = parserService.esTreeNodeToTSNodeMap.get(node);
        return typedTokenHelpers.getConstrainedTypeAtLocation(checker, tsNode);
    },
    expressionNodeIsArrayType(node, parserService, checker) {
        const nodeType = this.getNodeType(node, parserService, checker);
        return checker.isArrayType(nodeType);
    },
    getPropertyValueEqualsExpected(firstArgument, propertyName, expectedValue) {
        let didMatchExpectedValues = false;
        if (firstArgument !== undefined) {
            const foundPropertyOfName = firstArgument.properties.find((p) => p.type === TSESTree.AST_NODE_TYPES.Property &&
                p.key.name === propertyName);
            didMatchExpectedValues =
                foundPropertyOfName !== undefined &&
                    foundPropertyOfName.type === TSESTree.AST_NODE_TYPES.Property &&
                    foundPropertyOfName.value.value ===
                        expectedValue;
        }
        return didMatchExpectedValues;
    },
    getConstrainedTypeAtLocation(checker, node) {
        const nodeType = checker.getTypeAtLocation(node);
        const constrained = checker.getBaseConstraintOfType(nodeType);
        return constrained ?? nodeType;
    },
    nodeHasDecoratorsNamed(n, decoratorNames) {
        const decorators = this.getDecoratorsNamed(n, decoratorNames);
        return decorators.length > 0;
    },
    getDecoratorsNamed(n, decoratorNames) {
        const decorators = n.decorators.filter((d) => {
            const factoryMethodDecoratorIdentifier = d.expression
                .callee?.name;
            const decoratorIdentifier = d.expression
                .name;
            return decoratorNames.includes(factoryMethodDecoratorIdentifier ?? decoratorIdentifier ?? "");
        });
        return decorators || [];
    },
    parseStringToAst(code, path, context) {
        return parse(code, {
            filePath: path,
            range: true,
            tokens: true,
            loc: true,
            ...context.parserOptions,
        });
    },
    isEnumType(type) {
        // if for some reason this returns true...
        if (tsutils.isTypeFlagSet(type, ts.TypeFlags.Enum))
            return true;
        if (tsutils.isTypeFlagSet(type, ts.TypeFlags.EnumLike))
            return true;
        // it's not an enum type if it's an enum literal type
        if (tsutils.isTypeFlagSet(type, ts.TypeFlags.EnumLiteral) &&
            !type.isUnion())
            return false;
        // get the symbol and check if its value declaration is an enum declaration
        const symbol = type.getSymbol();
        if (symbol == null)
            return false;
        const { valueDeclaration } = symbol;
        return (valueDeclaration != null &&
            valueDeclaration.kind === ts.SyntaxKind.EnumDeclaration);
    },
    isOptionalPropertyValue(node) {
        const isUndefinedType = node.typeAnnotation?.typeAnnotation?.types?.find((t) => t.type === AST_NODE_TYPES.TSUndefinedKeyword) !== undefined;
        const isOptionalPropertyValue = node.optional || isUndefinedType || false;
        return isOptionalPropertyValue;
    },
    /**
     * Checks if an import is an import of the given decorator name
     * @param imp
     * @param decoratorName
     */
    importIsDecorator(imp, decoratorName) {
        const isFromClassValidator = imp.source.value.startsWith("class-validator");
        const isDecoratorImport = imp.specifiers.some((specifier) => specifier.local.name === decoratorName);
        return isFromClassValidator && isDecoratorImport;
    },
    /**
     * Checks if decorator is in imports of a node
     * @param imports
     * @param decorator
     */
    decoratorIsImportedFromClassValidator(imports, decorator) {
        const decoratorName = this.getDecoratorName(decorator);
        if (!decoratorName) {
            return false;
        }
        return imports.some((imp) => typedTokenHelpers.importIsDecorator(imp, decoratorName));
    },
    /**
     * Checks whether a decorator is a class validator decorator
     * @param program The root program node
     * @param decorator The decorator node
     */
    decoratorIsClassValidatorDecorator(program, decorator) {
        if (!program) {
            return false;
        }
        const imports = program.body.filter((node) => node.type === TSESTree.AST_NODE_TYPES.ImportDeclaration);
        return typedTokenHelpers.decoratorIsImportedFromClassValidator(imports, decorator);
    },
    /**
     * Gets the root program of a node
     * @param node
     */
    getRootProgram(node) {
        let root = node;
        while (root.parent) {
            if (root.parent.type === TSESTree.AST_NODE_TYPES.Program) {
                return root.parent;
            }
            root = root.parent;
        }
        return null;
    },
    /**
     * Gets all the decorators actually imported from class-validator lib or decorators that were included in the additionalCustomValidatorDecorators options
     * @param node PropertyDefinition node
     */
    getValidationDecorators(node, additionalCustomValidatorDecorators = []) {
        const program = typedTokenHelpers.getRootProgram(node);
        const { decorators } = node;
        return (decorators.filter((decorator) => {
            const isClassValidatorDecorator = typedTokenHelpers.decoratorIsClassValidatorDecorator(program, decorator);
            const decoratorName = typedTokenHelpers.getDecoratorName(decorator);
            const isCustomClassValidatorDecorator = decoratorName === null
                ? false
                : additionalCustomValidatorDecorators.includes(decoratorName);
            return (isCustomClassValidatorDecorator || isClassValidatorDecorator);
        }) ?? []);
    },
    /**
     * Checks if the decorator is the IsEnum decorator
     * @param decorator
     */
    decoratorIsIsEnum(decorator) {
        const decoratorName = this.getDecoratorName(decorator);
        return decoratorName === "IsEnum";
    },
    /** Checks if the decorator is the IsObject decorator
     * @param decorator
     */
    decoratorIsIsObject(decorator) {
        const decoratorName = this.getDecoratorName(decorator);
        return decoratorName === "IsObject";
    },
    /**
     * Gets the name of a decorator
     * Returns null if no name is found
     * @param decorator
     */
    getDecoratorName(decorator) {
        if (decorator.expression.type !== TSESTree.AST_NODE_TYPES.CallExpression) {
            return null;
        }
        if (decorator.expression.callee.type !==
            TSESTree.AST_NODE_TYPES.Identifier) {
            return null;
        }
        return decorator.expression.callee.name;
    },
};
