// @darraghor/eslint-plugin-nestjs-typed/dist/rules/allPropertiesAreWhitelisted/allPropertiesAreWhitelisted.js
import { AST_NODE_TYPES } from "@typescript-eslint/utils";
import { createRule } from "../../utils/createRule.js";
import { typedTokenHelpers } from "../../utils/typedTokenHelpers.js";
const rule = createRule({
    name: "all-properties-are-whitelisted",
    meta: {
        docs: {
            description: "Enforce all properties are whitelisted",
        },
        messages: {
            "missing-property-decorator": "Property has no class-validator decorator (use @Allow() if you don't need a validation)",
        },
        type: "problem",
        schema: [],
    },
    defaultOptions: [],
    create: function (context) {
        return {
            ClassDeclaration(node) {
                const program = typedTokenHelpers.getRootProgram(node);
                const withDecorator = [];
                const withoutDecorator = [];
                for (const element of node.body.body) {
                    if (element.type !== AST_NODE_TYPES.PropertyDefinition) {
                        continue;
                    }
                    const hasDecorator = element.decorators.some((decorator) => decorator.expression.type ===
                        AST_NODE_TYPES.CallExpression &&
                        decorator.expression.callee.type ===
                            AST_NODE_TYPES.Identifier &&
                        typedTokenHelpers.decoratorIsClassValidatorDecorator(program, decorator));
                    if (hasDecorator) {
                        withDecorator.push(element);
                    }
                    else {
                        withoutDecorator.push(element);
                    }
                }
                if (withDecorator.length > 0 && withoutDecorator.length > 0) {
                    for (const element of withoutDecorator) {
                        context.report({
                            node: element,
                            messageId: "missing-property-decorator",
                        });
                    }
                }
            },
        };
    },
});
export default rule;
