// @darraghor/eslint-plugin-nestjs-typed/dist/index.js
import noSwagger from "./configs/noSwagger.js";
import recommended from "./configs/recommended.js";
import rules from "./rules/index.js";
import * as parserBase from "@typescript-eslint/parser";
import { fileURLToPath } from "url";
import { readFileSync } from "fs";
import { dirname, resolve } from "path";
// eslint-disable-next-line @typescript-eslint/naming-convention
const __filename = fileURLToPath(import.meta.url);
// eslint-disable-next-line @typescript-eslint/naming-convention
const __dirname = dirname(__filename);
const { name, version } = JSON.parse(readFileSync(resolve(__dirname, "../package.json"), "utf8"));
// Most of this is copied and simplified from https://github.com/typescript-eslint/typescript-eslint/blob/v8.22.0/packages/typescript-eslint/src/configs/recommended.ts
// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
const meta = {
    name,
    version,
};
export const parser = {
    meta: parserBase.meta,
    parseForESLint: parserBase.parseForESLint,
};
const classicPlugin = {
    configs: {
        recommended: recommended,
        "no-swagger": noSwagger,
    },
    rules,
    meta,
};
export const plugin = classicPlugin;
const flatBaseConfig = (plugin, parser) => {
    const baseConfig = {
        name: "@darraghor/nestjs-typed/base",
        languageOptions: {
            parser,
            sourceType: "module",
        },
        plugins: {
            "@darraghor/nestjs-typed": plugin,
        },
    };
    return baseConfig;
};
// export the classic plugin configs
export { classicPlugin };
// export the flat configs
export default {
    plugin,
    configs: {
        flatRecommended: [
            flatBaseConfig(plugin, parser),
            {
                name: "@darraghor/nestjs-typed/recommended",
                rules: recommended.rules,
            },
        ],
        flatNoSwagger: [
            {
                name: "@darraghor/nestjs-typed/no-swagger",
                rules: noSwagger.rules,
            },
        ],
    },
};
