// @darraghor/eslint-plugin-nestjs-typed/dist/rules/apiPropertyMatchesPropertyOptionality/apiPropertyMatchesPropertyOptionality.js
import { createRule } from "../../utils/createRule.js";
import { typedTokenHelpers } from "../../utils/typedTokenHelpers.js";
export const shouldUseRequiredDecorator = (node) => {
    const hasOptionalDecorator = typedTokenHelpers.nodeHasDecoratorsNamed(node, ["ApiPropertyOptional"]);
    const isOptionalPropertyValue = typedTokenHelpers.isOptionalPropertyValue(node);
    return hasOptionalDecorator && !isOptionalPropertyValue;
};
export const shouldUseOptionalDecorator = (node) => {
    const hasRequiredDecorator = typedTokenHelpers.nodeHasDecoratorsNamed(node, ["ApiProperty"]);
    const isOptionalPropertyValue = typedTokenHelpers.isOptionalPropertyValue(node);
    return hasRequiredDecorator && isOptionalPropertyValue;
};
const rule = createRule({
    name: "api-property-matches-property-optionality",
    meta: {
        docs: {
            description: "Properties should have correct @ApiProperty decorators",
        },
        messages: {
            shouldUseOptionalDecorator: `Property marked as optional should use @ApiPropertyOptional decorator`,
            shouldUseRequiredDecorator: `Property marked as required should use @ApiProperty decorator`,
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        return {
            PropertyDefinition(node) {
                if (shouldUseOptionalDecorator(node)) {
                    context.report({
                        node: node,
                        messageId: "shouldUseOptionalDecorator",
                    });
                }
                if (shouldUseRequiredDecorator(node)) {
                    context.report({
                        node: node,
                        messageId: "shouldUseRequiredDecorator",
                    });
                }
            },
        };
    },
});
export default rule;
