// @darraghor/eslint-plugin-nestjs-typed/dist/rules/paramDecoratorNameMatchesRouteParam/paramDecoratorNameMatchesRouteParam.js
/* eslint-disable unicorn/prevent-abbreviations */
import { TSESTree } from "@typescript-eslint/utils";
import { createRule } from "../../utils/createRule.js";
const nestRequestMethodDecoratorNames = new Set([
    "Get",
    "Post",
    "Put",
    "Delete",
    "Patch",
    "Options",
    "Head",
    "All",
    "Sse",
]);
export const parsePathParts = (decorator) => {
    const decoratorArgument = decorator?.expression
        ?.arguments[0];
    if (decoratorArgument?.type === TSESTree.AST_NODE_TYPES.TemplateLiteral ||
        decoratorArgument?.type === TSESTree.AST_NODE_TYPES.Identifier) {
        return ["dareslint__skip"];
    }
    if (decoratorArgument?.type === TSESTree.AST_NODE_TYPES.Literal) {
        return [decoratorArgument.raw];
    }
    if (decoratorArgument?.type === TSESTree.AST_NODE_TYPES.ArrayExpression) {
        return decoratorArgument.elements.map((x) => x.raw);
    }
    if (decoratorArgument?.type === TSESTree.AST_NODE_TYPES.ObjectExpression) {
        return decoratorArgument.properties
            .filter((x) => x.key
            .name === "path")
            .map((x) => x.value.raw);
    }
    return [];
};
/**
 * nestjs allows for paths with _+?()*
 * this rule doesn't support parsing those so we'll just pass
 */
export const hasPathPartsAnyRegexParams = (pathPartsToCheck) => {
    // prettier-ignore
    const specialCharacterRegex = /(dareslint__skip|\*|\+|\?|\(|\)|_)/; //new RegExp("([\?\+\*\_\(\)])")
    return pathPartsToCheck.some((pathPart) => {
        return specialCharacterRegex.test(pathPart);
    });
};
/**
 * Checks if there is a matching path part for the paramName
 * @param paramName
 * @param pathPartsToCheck
 * @returns
 */
export const isParameterNameIncludedInAPathPart = (paramName, pathPartsToCheck) => {
    return pathPartsToCheck.some((pathPart) => {
        return (
        // note to reader: this might be better as a regex. feel free to open a pr!
        pathPart === `":${paramName}"` ||
            pathPart === `':${paramName}'` ||
            pathPart.includes(`/:${paramName}/`) ||
            pathPart.includes(`/:${paramName}"`) ||
            pathPart.includes(`":${paramName}/`) ||
            pathPart.includes(`/:${paramName}'`) ||
            pathPart.includes(`':${paramName}/`));
    });
};
export const shouldTrigger = (decorator, ruleOptions) => {
    if (!decorator) {
        return {
            hasColonInName: false,
            paramNameNotMatchedInPath: false,
        };
    }
    // grab the param name
    const paramName = (decorator?.expression)
        .arguments[0]?.value;
    // if there's no param name get out of here
    if (!paramName || paramName === "") {
        return {
            hasColonInName: false,
            paramNameNotMatchedInPath: false,
        };
    }
    // param names don't need the colon
    if (paramName.startsWith(":")) {
        return {
            hasColonInName: true,
            paramNameNotMatchedInPath: false,
        };
    }
    let pathPartsToCheck = [];
    // grab any controller path parts
    if (ruleOptions[0].shouldCheckController) {
        // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
        const controllerDecorator = (decorator.parent.parent?.parent?.parent
            ?.parent).decorators.find((d) => {
            return (d.expression
                .callee.name === "Controller");
        });
        pathPartsToCheck = pathPartsToCheck.concat(parsePathParts(controllerDecorator));
    }
    // grab any api method path parts from method decorator
    const methodDefinition = decorator.parent.parent
        ?.parent;
    // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
    const methodDecorator = methodDefinition.decorators.find((d) => {
        return nestRequestMethodDecoratorNames.has(d.expression
            .callee.name);
    });
    pathPartsToCheck = pathPartsToCheck.concat(parsePathParts(methodDecorator));
    const shouldIgnoreThisSetOfRoutes = 
    // is a template literal argument
    // is an identifier argument
    hasPathPartsAnyRegexParams(pathPartsToCheck);
    if (shouldIgnoreThisSetOfRoutes) {
        return {
            hasColonInName: false,
            paramNameNotMatchedInPath: false,
        };
    }
    // check that the param name is in one path part
    return {
        hasColonInName: false,
        paramNameNotMatchedInPath: !isParameterNameIncludedInAPathPart(paramName, pathPartsToCheck),
        paramName,
    };
};
const defaultRuleOptions = [{ shouldCheckController: true }];
const rule = createRule({
    name: "param-decorator-name-matches-route-param",
    meta: {
        docs: {
            description: 'Param decorators with a name parameter e.g. Param("myvar") should match a specified route parameter - e.g. Get(":myvar")',
        },
        messages: {
            paramIdentifierDoesntNeedColon: "You don't need to specify the colon (:) in a Param decorator",
            paramIdentifierShouldMatchRouteOrController: 'The Param "{{paramName}}" could not be found in the related route or controller',
            paramIdentifierShouldMatchRouteOnly: 'The Param "{{paramName}}" could not be found in the related route',
        },
        schema: [
            {
                type: "object",
                properties: {
                    shouldCheckController: {
                        description: "If the name in the @Controller() decorator should be checked for route param matches or not. Turn this option off if you use variable for Controller paths that do not contain route params.",
                        type: "boolean",
                    },
                },
            },
        ],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: defaultRuleOptions,
    create(contextWithoutDefaults) {
        const context = contextWithoutDefaults.options &&
            contextWithoutDefaults.options.length > 0
            ? contextWithoutDefaults
            : // only apply the defaults when the user provides no config
                Object.setPrototypeOf({
                    options: defaultRuleOptions,
                }, contextWithoutDefaults);
        return {
            Decorator(node) {
                if (node.expression
                    .callee.name !== "Param") {
                    return;
                }
                const result = shouldTrigger(node, context.options);
                if (result.paramNameNotMatchedInPath) {
                    context.report({
                        node: node,
                        messageId: context.options[0].shouldCheckController
                            ? "paramIdentifierShouldMatchRouteOrController"
                            : "paramIdentifierShouldMatchRouteOnly",
                        data: { paramName: result.paramName },
                    });
                }
                if (result.hasColonInName) {
                    context.report({
                        node: node,
                        messageId: "paramIdentifierDoesntNeedColon",
                    });
                }
            },
        };
    },
});
export default rule;
