// @darraghor/eslint-plugin-nestjs-typed/dist/utils/files/customFileEnumeratorWrapper.js
import { glob } from "glob";
import path from "path";
import IsFilteredPath from "./isFilteredPath.js";
// eslint-disable-next-line @typescript-eslint/no-extraneous-class
export class FileEnumerator {
    static enumerateFiles({ sourceGlobs, extensions, filterFromPaths, }) {
        // Create glob pattern that includes extensions
        const patterns = sourceGlobs.flatMap((filePath) => {
            if (path.extname(filePath)) {
                return filePath;
            }
            return extensions.map((extension) => `${filePath}/**/*${extension}`);
        });
        // Find all files matching the patterns
        const files = glob.sync(patterns, {
            ignore: ["**/node_modules/**"],
            nodir: true,
            absolute: true,
        });
        // Map to the expected format and filter
        return files
            .map((filePath) => ({
            ignored: false, // You can implement custom ignore logic if needed
            filename: filePath,
        }))
            .filter((file) => {
            const isFiltered = IsFilteredPath.test(file.filename, filterFromPaths);
            return !isFiltered;
        });
    }
}
export default FileEnumerator;
